Implement hstack, vstack, dstack

muthuArivoli opened the pull request:
Related to #38349

muthuArivoli committed:
implement hstack, vstack, dstack

muthuArivoli committed:
add test

muthuArivoli committed:
add docs

muthuArivoli committed:
update tests

muthuArivoli committed:
change single tensor to list of single tensor

muthuArivoli committed:
fix test

muthuArivoli committed:
reduce tests

muthuArivoli committed:
catch empty tensorlist error

muthuArivoli commented:
@mruberry PTAL

mruberry commented on torch/_torch_docs.py:
```diff
@@ -1016,6 +1016,100 @@ def merge_dicts(*dicts):
+    tensor([[1, 2, 3],
+            [2, 3, 4]])
+    >>> a = torch.tensor([[1],[2],[3]])
+    >>> b = torch.tensor([[2],[3],[4]])
```
Nit: these examples are excellent but maybe

```
a = torch.tensor([[1],[2],[3]])
b = torch.tensor([[4],[5],[6]])
```

would be clearer?

mruberry commented on torch/_torch_docs.py:
```diff
@@ -1016,6 +1016,100 @@ def merge_dicts(*dicts):
+Example::
+
+    >>> a = torch.tensor([1, 2, 3])
+    >>> b = torch.tensor([2, 3, 4])
```
See numbering suggestion below.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -1016,6 +1016,100 @@ def merge_dicts(*dicts):
+    >>> a = torch.tensor([1, 2, 3])
+    >>> b = torch.tensor([2, 3, 4])
+    >>> torch.dstack((a,b))
+    tensor([[[1, 2],
```
See numbering suggestion above.

mruberry commented on test/test_torch.py:
```diff
@@ -18433,6 +18433,35 @@ def test_atleast(self, device, dtype):
         self._test_atleast_dim(torch.atleast_2d, np.atleast_2d, device, dtype)
         self._test_atleast_dim(torch.atleast_3d, np.atleast_3d, device, dtype)
 
+    def _test_special_stacks(self, torch_fn, np_fn, device, dtype):
```
These tests are good, but this test's case generation is limited to replicating the same tensor shape for each element of the input list. Here are some cases I was thinking about:

- op(t) 
  - the behavior of `np.hstack(a)` is strange and `np.hstack(a) != np.hstack((a,))` (the same is true for `np.dstack`)
  - do we even support non-tuple arguments? if not we should validate this throws a runtime error
  - if we support single tensor arguments, is `np.hstack(a)`'s and `np.dstack(a)`'s behavior correct? 
- op((a, b, c, ...))
  - validating that if they differ on an unexpected dim an error is thrown (maybe `_test_special_stacks` should take a dim argument corresponding to the op?)
  - validating that if they differ only on the expected dim the result is equivalent to NumPy
-  are tensors with a size zero dim handled correctly? (if not that's OK, but let's assert it doesn't work)
-  `np.hstack` has special-handling of 1D tensors (as your implementation does), does `test_hstack` need a custom elaboration to test that behavior, in particular?
- validating that tensors with different shapes but the same post-`atleast_Xd` shapes meet the criteria work

For an example of the last bullet:

```
a = np.array([[[1],[2],[3]]])
b = np.array((4, 5, 6))
np.dstack((a, b))
: array([[[1, 4],
          [2, 5],
          [3, 6]]])
```

This is a good number of cases but validating each one by hand shouldn't be too laborious, I hope. 

What are your thoughts? Are there other cases I missed?

mruberry reviewed (changes requested):
Overall looks excellent. A couple minor nits about the doc examples and questions about test coverage.

muthuArivoli committed:
rewrite tests

muthuArivoli committed:
fix type of error

muthuArivoli committed:
add at_least_dim param to tests

muthuArivoli committed:
Merge remote-tracking branch 'upstream/master' into implement-stack

muthuArivoli committed:
fix loop variable and at least dim

muthuArivoli committed:
add better scalar test and fix tuples

muthuArivoli committed:
fix tuple and add autograd tests

muthuArivoli committed:
fix xla failure

muthuArivoli committed:
Merge remote-tracking branch 'upstream/master' into implement-stack

muthuArivoli commented:
@mruberry I have added tests that I think cover all of the cases. They cover:
* Runtime Error for non-tuple argument
* 0 dimension support
* For vstack and dstack, test with tensors with same post at_least shape, but different original shape 
* Test whether the function is accurate when the only dim that is different is the dim that will be concatenated
* Test whether a runtime error is thrown when the changed dim is different than the one that is concatenated

For the last two, those are tested from 1 to 4 dimensions, so the special behavior for hstack is included with that. I have also added some autograd tests in a similar manner to the existing stack autograd test.

Does this sound good, or do I need more tests?

mruberry commented on test/test_torch.py:
```diff
@@ -18664,6 +18664,105 @@ def test_atleast(self, device, dtype):
+                else:
+                    # Invalid dimensions, test for error
+                    with self.assertRaisesRegex(RuntimeError, "Sizes of tensors must match except in dimension"):
+                        torch_fn(torch_input)
```
Would you add an assert that NumPy also throws a runtime error in this case? You don't need to assert a string is thrown:

```
with self.assertRaises(RuntimeError):
  np_fn(np_input)
```

mruberry reviewed (approved):
Nice work, @muthuArivoli! 

Would you just fix that one minor nit on the tests and we'll get this merged?

Let me know if you're interested in working on a new problem.

muthuArivoli committed:
add numpy error

muthuArivoli commented:
@mruberry I added the numpy error check. Is it ok that numpy throws a ValueError, while we throw a RuntimeError?

Yes, I'm interested in working on a new problem, do you have any recommendations?

mruberry commented:
> @mruberry I added the numpy error check. Is it ok that numpy throws a ValueError, while we throw a RuntimeError?

Absolutely OK. Nice work.

> Yes, I'm interested in working on a new problem, do you have any recommendations?

For symmetry there are the split functions, hsplit, vsplit, and dsplit. A slightly more challenging binary function is divmod, because it returns two tensors. There are the "polynomial" functions, like polyadd and polyder, but I'm hoping someone will write all of them near simultaneously because they have a lot of common structure. There are also unary functions, like nan_to_num, that would be very helpful. 

If you'd like something more exotic or especially numerically challenging there are also functions like the kaiser windowing function.

muthuArivoli commented:
Two questions:

1. For hsplit, vsplit, dsplit, do the functions need to be exactly like their numpy equivalents, or can they just wrap around the torch.split function? The torch.split function is different than the np.split function when the second argument is a list, with np using indices and torch using lengths, but they essentially perform the same role.

2. For the poly functions, it seems like the ones linked in the tracking issue have been superseded by [np.polynomial](https://numpy.org/doc/stable/reference/routines.polynomials.polynomial.html), and that the main reason they are still around are for BC reasons. Should the implementation follow the functions linked in the tracking issue, or the ones linked here, and would a new namespace need to be created? Mainly asking since the representation of the degrees of the polynomial was reversed between the two.

mruberry commented:
> Two questions:
> 
> 1. For hsplit, vsplit, dsplit, do the functions need to be exactly like their numpy equivalents, or can they just wrap around the torch.split function? The torch.split function is different than the np.split function when the second argument is a list, with np using indices and torch using lengths, but they essentially perform the same role.
> 2. For the poly functions, it seems like the ones linked in the tracking issue have been superseded by [np.polynomial](https://numpy.org/doc/stable/reference/routines.polynomials.polynomial.html), and that the main reason they are still around are for BC reasons. Should the implementation follow the functions linked in the tracking issue, or the ones linked here, and would a new namespace need to be created? Mainly asking since the representation of the degrees of the polynomial was reversed between the two.

Excellent questions. 

1) Analogous to torch.split's behavior.
2) You're correct and I should update the references to point to the newest ones in the new namespace, and you're right that would require adding a new namespace, which is kinda a pain to do. I should probably remove those until we decide to implement the polynomial namespace.
